src/Serializers/Orleans.Serialization.Protobuf/ProtobufCodec.cs

```diff
@@ -0,0 +1,210 @@
+using Orleans.Serialization.WireProtocol;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace Orleans.Serialization;
+
+[Alias(WellKnownAlias)]
+public sealed class ProtobufCodec : IGeneralizedCodec, IGeneralizedCopier, ITypeFilter
+{
+    public const string WellKnownAlias = "protobuf";
+
+    private static readonly ConcurrentDictionary<RuntimeTypeHandle, MessageParser> Parsers = new();
+
+    private readonly ICodecSelector[] _serializableTypeSelectors;
+    private readonly ICopierSelector[] _copyableTypeSelectors;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ProtobufCodec"/> class.
+    /// </summary>
+    /// <param name="serializableTypeSelectors">Filters used to indicate which types should be serialized by this codec.</param>
+    /// <param name="copyableTypeSelectors">Filters used to indicate which types should be copied by this codec.</param>
+    public ProtobufCodec(
+        IEnumerable<ICodecSelector> serializableTypeSelectors,
+        IEnumerable<ICopierSelector> copyableTypeSelectors)
+    {
+        _serializableTypeSelectors = serializableTypeSelectors.Where(t => string.Equals(t.CodecName, WellKnownAlias, StringComparison.Ordinal)).ToArray();
+        _copyableTypeSelectors = copyableTypeSelectors.Where(t => string.Equals(t.CopierName, WellKnownAlias, StringComparison.Ordinal)).ToArray();
+    }
+
+    public object DeepCopy(object input, CopyContext context)
+    {
+        if (!context.TryGetCopy(input, out object result))
+        {
+            dynamic dynamicSource = input;
+            result = dynamicSource.Clone();
+
+            context.RecordCopy(input, result);
+        }
+
+        return result;
+    }
+
+    /// <inheritdoc/>
+    bool IGeneralizedCodec.IsSupportedType(Type type)
+    {
+        foreach (var selector in _serializableTypeSelectors)
+        {
+            if (selector.IsSupportedType(type))
+            {
+                return IsMessageParser(type);
+            }
+        }
+
+        return false;
+    }
+
+    /// <inheritdoc/>
+    bool IGeneralizedCopier.IsSupportedType(Type type)
+    {
+        foreach (var selector in _copyableTypeSelectors)
+        {
+            if (selector.IsSupportedType(type))
+            {
+                return IsMessageParser(type);
+            }
+        }
+
+        return false;
+    }
+
+    /// <inheritdoc/>
+    bool? ITypeFilter.IsTypeAllowed(Type type)
+    {
+        if (!typeof(IMessage).IsAssignableFrom(type))
+        {
+            return null;
+        }
+
+        return ((IGeneralizedCodec)this).IsSupportedType(type) || ((IGeneralizedCopier)this).IsSupportedType(type);
+    } 
+
+    static bool IsMessageParser(Type type)
+    {
+        if (!Parsers.ContainsKey(type.TypeHandle))
+        {
+            var prop = type.GetProperty("Parser", BindingFlags.Public | BindingFlags.Static);
```

Comment:
Is there a way to avoid reflection here?

Reply:
We either need a reference to the static `Parser` prop (as shown here using reflection), or we can create an instance from the `Type` using `System.Activator.CreateInstance(type)` to access the parser (latest commit). Not sure about other possibilities to be honest. What do you think?

Reply:
Yeah, it would be nice if we could get the protobuf team to add a static `Parser` property to the `IMessage` interface... But even then if all you have is a `Type` object, and we want to cache the `Parser` for later use, I'm not sure how to get at it without reflection. `Activator.CreateInstance` is going to use reflection to get the ctor anyway 🤷

Reply:
If we could add a static interface method to IMessage for the parser, that might solve the problem. Maybe the Protobuf folks would be open to it.

For now, this is acceptable